executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+            }
+            InterpreterAction::AssignExpression(idx, _) => {
+                set.insert(*idx);
+            }
+            InterpreterAction::MachineCall(_, params) => params.iter().for_each(|p| {
+                if let MachineCallArgumentIdx::Unknown(v) = p {
+                    set.insert(*v);
+                }
+            }),
+            _ => {}
+        }
+        set
+    }
+
+    // variable indexes read by the action
+    fn reads(&self) -> BTreeSet<usize> {
+        let mut set = BTreeSet::new();
+        match self {
+            InterpreterAction::WriteCell(idx, _) => {
+                set.insert(*idx);
+            }
+            InterpreterAction::WriteParam(idx, _) => {
+                set.insert(*idx);
+            }
+            InterpreterAction::AssignExpression(_, expr) => expr.elems.iter().for_each(|e| {
+                if let RPNExpressionElem::Symbol(idx) = e {
+                    set.insert(*idx);
+                }
+            }),
+            InterpreterAction::MachineCall(_, params) => params.iter().for_each(|p| {
+                if let MachineCallArgumentIdx::Known(v) = p {
+                    set.insert(*v);
+                }
+            }),
+            InterpreterAction::Assertion(lhs, rhs, _) => {
+                lhs.elems.iter().for_each(|e| {
+                    if let RPNExpressionElem::Symbol(idx) = e {
+                        set.insert(*idx);
+                    }
+                });
+                rhs.elems.iter().for_each(|e| {
+                    if let RPNExpressionElem::Symbol(idx) = e {
+                        set.insert(*idx);
+                    }
+                });
+            }
+            _ => {}
+        }
+        set
+    }
+}
+
+/// Helper struct to map variables to unique indices, so they can be kept in
+/// sequential memory and quickly refered to during execution.
+pub struct VariableMapper {
+    var_idx: HashMap<Variable, usize>,
+    count: usize,
+}
+
+impl VariableMapper {
+    pub fn new() -> Self {
+        Self {
+            var_idx: HashMap::new(),
+            count: 0,
+        }
+    }
+
+    pub fn var_count(&self) -> usize {
+        self.count
+    }
+
+    pub fn map_var(&mut self, var: &Variable) -> usize {
+        let idx = *self.var_idx.entry(var.clone()).or_insert_with(|| {
+            self.count += 1;
+            self.count - 1
+        });
+        idx
+    }
+
+    // reserve a new variable index
+    pub fn reserve_idx(&mut self) -> usize {
+        let idx = self.count;
+        self.count += 1;
+        idx
+    }
+
+    // get the index of a variable if it was previously mapped
+    pub fn get_var(&mut self, var: &Variable) -> Option<usize> {
+        self.var_idx.get(var).copied()
+    }
+
+    pub fn map_expr_to_rpn<T: FieldElement>(
+        &mut self,
+        expr: &SymbolicExpression<T, Variable>,
+    ) -> RPNExpression<T, usize> {
+        RPNExpression::from(&expr.map_variables(&mut |var| self.map_var(var)))
```

Comment:
Does it make sense to pass (a mut ref to) the var mapper to the rpn conversion routine so that we only have to map the var for the symbol itself instead of creating a full copy of the expression?